File: opengl/texture/buffer/TextTextureBuffer.java
package org.anddev.andengine.opengl.texture.buffer;

using andengine.opengl.buffer.BufferObject;
using andengine.opengl.font.Font;
using andengine.opengl.font.Letter;
using andengine.opengl.util.FastFloatBuffer;

/**
 * @author Nicolas Gramlich
 * @since 11:05:56 - 03.04.2010
 */
public class TextTextureBuffer extends BufferObject {
	// ===========================================================
	// Constants
	// ===========================================================

	// ===========================================================
	// Fields
	// ===========================================================

	// ===========================================================
	// Constructors
	// ===========================================================

	public TextTextureBuffer(final int pCapacity, final int pDrawType) {
		super(pCapacity, pDrawType);
	}

	// ===========================================================
	// Getter & Setter
	// ===========================================================

	// ===========================================================
	// Methods for/from SuperClass/Interfaces
	// ===========================================================

	// ===========================================================
	// Methods
	// ===========================================================

	public synchronized void update(final Font pFont, final String[] pLines) {
		final FastFloatBuffer textureFloatBuffer = this.getFloatBuffer();
		textureFloatBuffer.position(0);

		final Font font = pFont;
		final String[] lines = pLines;

		final int lineCount = lines.length;
		for (int i = 0; i < lineCount; i++) {
			final String line = lines[i];

			final int lineLength = line.length();
			for (int j = 0; j < lineLength; j++) {
				final Letter letter = font.getLetter(line.charAt(j));

				final float letterTextureX = letter.mTextureX;
				final float letterTextureY = letter.mTextureY;
				final float letterTextureX2 = letterTextureX + letter.mTextureWidth;
				final float letterTextureY2 = letterTextureY + letter.mTextureHeight;

				textureFloatBuffer.put(letterTextureX);
				textureFloatBuffer.put(letterTextureY);

				textureFloatBuffer.put(letterTextureX);
				textureFloatBuffer.put(letterTextureY2);

				textureFloatBuffer.put(letterTextureX2);
				textureFloatBuffer.put(letterTextureY2);

				textureFloatBuffer.put(letterTextureX2);
				textureFloatBuffer.put(letterTextureY2);

				textureFloatBuffer.put(letterTextureX2);
				textureFloatBuffer.put(letterTextureY);

				textureFloatBuffer.put(letterTextureX);
				textureFloatBuffer.put(letterTextureY);
			}
		}
		textureFloatBuffer.position(0);

		this.setHardwareBufferNeedsUpdate();
	}

	// ===========================================================
	// Inner and Anonymous Classes
	// ===========================================================
}

File: opengl/texture/buffer/TextTextureBuffer.cs
namespace andengine.opengl.texture.buffer
{

    using BufferObject = andengine.opengl.buffer.BufferObject;
    using Font = andengine.opengl.font.Font;
    using Letter = andengine.opengl.font.Letter;
    using FastFloatBuffer = andengine.opengl.util.FastFloatBuffer;
    using System.Runtime.CompilerServices;
    using Java.Lang;

    /**
     * @author Nicolas Gramlich
     * @since 11:05:56 - 03.04.2010
     */
    public class TextTextureBuffer : BufferObject
    {
        // ===========================================================
        // Constants
        // ===========================================================

        // ===========================================================
        // Fields
        // ===========================================================

        // ===========================================================
        // Constructors
        // ===========================================================

        public TextTextureBuffer(int pCapacity, int pDrawType)
            : base(pCapacity, pDrawType)
        {
        }

        // ===========================================================
        // Getter & Setter
        // ===========================================================

        // ===========================================================
        // Methods for/from SuperClass/Interfaces
        // ===========================================================

        // ===========================================================
        // Methods
        // ===========================================================

        public void update(Font pFont, String[] pLines)
        {
            lock (_methodLock)
            {
                FastFloatBuffer textureFloatBuffer = this.GetFloatBuffer();
                textureFloatBuffer.Position(0);

                Font font = pFont;
                String[] lines = pLines;

                int lineCount = lines.Length;
                for (int i = 0; i < lineCount; i++)
                {
                    String line = lines[i];

                    int lineLength = line.Length();
                    for (int j = 0; j < lineLength; j++)
                    {
                        Letter letter = font.GetLetter(line.CharAt(j));

                        float letterTextureX = letter.mTextureX;
                        float letterTextureY = letter.mTextureY;
                        float letterTextureX2 = letterTextureX + letter.mTextureWidth;
                        float letterTextureY2 = letterTextureY + letter.mTextureHeight;

                        textureFloatBuffer.Put(letterTextureX);
                        textureFloatBuffer.Put(letterTextureY);

                        textureFloatBuffer.Put(letterTextureX);
                        textureFloatBuffer.Put(letterTextureY2);

                        textureFloatBuffer.Put(letterTextureX2);
                        textureFloatBuffer.Put(letterTextureY2);

                        textureFloatBuffer.Put(letterTextureX2);
                        textureFloatBuffer.Put(letterTextureY2);

                        textureFloatBuffer.Put(letterTextureX2);
                        textureFloatBuffer.Put(letterTextureY);

                        textureFloatBuffer.Put(letterTextureX);
                        textureFloatBuffer.Put(letterTextureY);
                    }
                }
                textureFloatBuffer.Position(0);

                this.SetHardwareBufferNeedsUpdate();
            }
        }

        // ===========================================================
        // Inner and Anonymous Classes
        // ===========================================================
    }
}